### PR TITLE
Added support for rectangular textures

### PR DESCRIPTION
**Pull request by thecodewarrior**
Adds support for non-square textures through the `width` and `height` mcmeta properties

Example image
All the sides in this image are the full texture, they aren't the bottom half of a square texture.

[Dinnerbone said that it existed](https://twitter.com/Dinnerbone/status/552241757497622529), and the properties are parsed in vanilla, but it isn't implemented. This fixes that. I'll create a bug report for minecraft 1.10 so they can implement it in vanilla.

This patch also adds an error if the height calculated based on the aspect ratio is taller than the image. It shows up just like the error if the image isn't square and doesn't have an MCMETA file.

## MCMETA properties

```
{
    "animation": {
        "width": [width component of aspect ratio, default 1],
        "height": [height component of aspect ratio, default 1],
        [everything else for your animation]
    }
}
```

### Static 2:1 texture  (Third block from the left)

Texture:
Texture

MCMETA:

```
{
    "animation": {
        "width": 2
    }
}
```

### Animated 4:3 texture (Fourth block from the right)

Texture:
Texture
w/ ruler and colored to show frame boundaries.

MCMETA:

```
{
    "animation": {
        "width": 4,
        "height": 3,
        "frames": [
            0, 1, 2, 3
        ]
    }
}
```

## Stitching

The only change I had to make with the stitching is disable the rotation of sprites that are taller than they are wide. Minecraft's stitcher will rotate sprites that're taller than they are wide to fit them more efficiently but won't rotate the textures when it writes them into the atlas or when interpolating UV coords. I suspect the stitcher was copied verbatim from somewhere else that included the rotation code, but minecraft doesn't support it.

Atlas

The blue highlights show where if a gap is less than 16x16 it won't be used, and the green highlight shows that a 16x16 gap can be utilized by the stitcher to fit other 16x textures.

## Other changes

I made a few changes that aren't strictly related to the PR, but were required in the testing process and may be useful in the future.

I added a debug mod based on [this](https://gist.github.com/LexManos/3c700306331080598daf) that will export the atlas whenever it's generated.

I also added a config option to forge that will fill the atlas with 50% transparency 0xFF00FF before generating it, allowing you to see exactly what is being written with a texture. That required a small patch to TextureUtil.

### CODE REVIEW

File: patches/minecraft/net/minecraft/client/renderer/texture/TextureAtlasSprite.java.patch [hunk 1]
```diff
@@ -17,7 +17,75 @@
...
+     public void func_188538_a(PngSizeInfo p_188538_1_, boolean p_188538_2_) throws IOException
+     {
++        loadSprite(p_188538_1_, p_188538_2_, null);
++    }
++    
++    public void loadSprite(PngSizeInfo sizeInfo, boolean p_188538_2_, IResource resource) throws IOException
++    {
+         this.func_130102_n();
+-        this.field_130223_c = p_188538_1_.field_188533_a;
+-        this.field_130224_d = p_188538_1_.field_188534_b;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/client/renderer/texture/TextureAtlasSprite.java.patch]: you need exc lines.

File: patches/minecraft/net/minecraft/client/renderer/texture/TextureUtil.java.patch [hunk 2]
```diff
@@ -1,6 +1,18 @@
...
-@@ -202,9 +202,11 @@
+@@ -12,8 +12,11 @@
+ import net.minecraft.client.resources.IResource;
+ import net.minecraft.client.resources.IResourceManager;
+ import net.minecraft.util.ResourceLocation;
++import net.minecraftforge.common.ForgeModContainer;
+ import net.minecraftforge.fml.relauncher.Side;
+ import net.minecraftforge.fml.relauncher.SideOnly;
++import scala.actors.threadpool.Arrays;
++
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/client/renderer/texture/TextureUtil.java.patch]: Why are you pilling in scala? Also why are you importing anything at all?

Comment (thecodewarrior) [inline patches/minecraft/net/minecraft/client/renderer/texture/TextureUtil.java.patch]: whoops. Stupid scala and automatic imports.

File: patches/minecraft/net/minecraft/client/renderer/texture/TextureAtlasSprite.java.patch (hunk 1 above)
Comment (thecodewarrior) [inline patches/minecraft/net/minecraft/client/renderer/texture/TextureAtlasSprite.java.patch]: What does exc lines mean?

**Comment by williewillus** [comment]
You should throw these examples on RTD after this gets pulled :D

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Pull request closed without merging**